api/src/main/java/org/apache/iceberg/catalog/SupportsCatalogTransactions.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * <p>Catalog implementations are not required to support catalog-level transactional state.
+ * If they do, they may support one or more {@code IsolationLevel}s and one or more
+ * {@code LockingMode}s.
+ */
+public interface SupportsCatalogTransactions {
+
+  /**
+   * The level of isolation for a catalog-level transaction.
+   *
+   * <p>Isolation covers both what data is read and what data can be written.
+   *
+   * <p>At all levels, data is only visible if it is either committed by another transaction or
+   * committed by a nested transaction within this catalog-level transaction.
+   *
+   * <p>Individual nested Table transactions may be "rebased" to expose updated versions of a
+   * table if the isolation level allows that behavior.
+   *
+   * <p>In the definitions of each isolation level, the concept of conflicting writes is
+   * referenced. Conflicting writes are two mutations to the same object that happen concurrently.
+   * Depending on the particular implementation, the coarseness of this conflict may vary. The
+   * most coarse conflict is any two mutations to the same table. However, some implementations
+   * may consider some of these "absolute" conflicts as allowable by using finer-grained conflict
+   * resolution. For example, two different operations that both append new files to a table may
+   * be in "absolute" conflict but could be resolved automatically as a "safe conflict" by using
+   * a set of automatic implementation-defined conflict resolution rules.
+   */
+  enum IsolationLevel {
+
+    /**
+     * Reading the same table multiple times may result in different versions read of the same
+     * table. A commit can be completed as long as any tables changed externally do not conflict
+     * with any writes within this transaction.
+     */
+    READ_COMMITTED,
+
+    /**
+     * Reading the same table multiple times will result in the same view of that table.
+     * Different tables may come from different snapshots. A commit can be completed as
+     * long as any tables changed externally do not conflict with any writes within this
+     * transaction.
+     */
+    REPEATED_READ,
+
+    /**
+     * A commit will only succeed if there have been no meaningful changes to data read during
+     * the course of this transaction prior to commit. This imposes stricter read guarantees than
+     * {@code REPEATED_READ} (consistent reads per table) as it requires that the reads are
+     * consistent for all tables to a single point in time (or single snapshot of the database).
+     * Additionally, it implies additional requirements around the successful completion of a
+     * write. In order for a write to complete, any entities read during this transaction are also
+     * blocked from changing (via another transaction) post-read in ways that would influence the
+     * writes of this operation. This is also sometimes called snapshot isolation.
```

Comment:
Here is my understanding: because commits of each table must be serializable, and each multi-table commit is a series of single-table commits (no commit operation affecting multiple tables), there is no need for a range lock. Because of that, locking all tables before commit (serializable) and checking each table version at commit time (snapshot isolation) is equivalent, as long as the reader only sees the data of the same snapshot of all tables during the transaction.

Reply:
Snapshot isolation in general is more sticky from my pov as I don't believe there is a canonical definition of it. Serializable has a very clear definition from sql 92. People were confused previously by this definition and the missing snapshot isolation which is what caused me to add this sentence.

Reply:
I updated the sentence here to clarify that it covers those guarantees, rather than suggesting they are interchangabel.